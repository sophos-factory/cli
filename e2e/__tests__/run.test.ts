/* eslint-disable @typescript-eslint/no-non-null-assertion */
import isArray from 'lodash/isArray';
import isNil from 'lodash/isNil';
import { execute } from '../helpers/execute';
import knownIds from '../helpers/knownIds';

describe('refactrctl run', () => {
  // Give platform time to bootstrap runner.
  jest.setTimeout(60 * 1000);
  describe('pipeline', () => {
    test('throws on missing arguments', async () => {
      await expect(
        execute(['run', 'pipeline'], {
          token: process.env.FACTORY_DYNAMIC_AUTH_TOKEN!
        })
      ).rejects.toMatchSnapshot();
    });

    test('returns run object without --wait flag', async () => {
      await expect(
        execute(
          [
            'run',
            'pipeline',
            '--project-id',
            knownIds.dynamic.project,
            '--revision-id',
            knownIds.dynamic.pipelineRevision,
            '--var',
            '\'string_array_variable:["string", "array"]\'',
            '--var',
            '\'string_variable:"string"\'',
            '--var',
            "'number_array_variable:[100, 123]'",
            '--var',
            "'number_variable:123'",
            '--var',
            "'boolean_variable:true'",
            knownIds.dynamic.pipeline,
            '--format=json'
          ],
          { token: process.env.FACTORY_DYNAMIC_AUTH_TOKEN! }
        ).then((value) => JSON.parse(value))
      ).resolves.toHaveProperty('status', 'Queued');
    });

    test('returns an array of events with --wait flag', async () => {
      await expect(
        execute(
          [
            'run',
            'pipeline',
            '--project-id',
            knownIds.dynamic.project,
            '--revision-id',
            knownIds.dynamic.pipelineRevision,
            knownIds.dynamic.pipeline,
            '--wait',
            '--format=json'
          ],
          { token: process.env.FACTORY_DYNAMIC_AUTH_TOKEN! }
        ).then((value) => isArray(JSON.parse(value)))
      ).resolves.toBeTruthy();
    });

    test('correctly runs with variables', async () => {
      await expect(
        (async () => {
          const result = JSON.parse(
            await execute(
              [
                'run',
                'pipeline',
                '--project-id',
                knownIds.dynamic.project,
                '--revision-id',
                knownIds.dynamic.pipelineRevision,
                '--var',
                '\'string_array_variable:["string", "array"]\'',
                '--var',
                '\'string_variable:"string"\'',
                '--var',
                "'number_array_variable:[100, 123]'",
                '--var',
                "'number_variable:123'",
                '--var',
                "'boolean_variable:true'",
                knownIds.dynamic.pipeline,
                '--wait',
                '--format=json'
              ],
              { token: process.env.FACTORY_DYNAMIC_AUTH_TOKEN! }
            )
          );

          if (!isArray(result)) {
            throw new Error('Expected array to be received!');
          }

          const ret = result.find(
            (entry) =>
              entry.message === 'Debug message' && entry.code === 'StepOutput'
          );
          if (isNil(ret)) {
            throw new Error('Expected debug step output to be presented!');
          }

          return JSON.parse(ret.details);
        })()
      ).resolves.toStrictEqual({
        string_array_variable: ['string', 'array'],
        string_variable: 'string',
        number_array_variable: [100, 123],
        number_variable: 123,
        boolean_variable: true
      });
    });
  });

  describe('with --wait', () => {
    let runId: string;
    beforeAll(async () => {
      const data = JSON.parse(
        await execute(
          [
            'run',
            'pipeline',
            '--project-id',
            knownIds.dynamic.project,
            '--revision-id',
            knownIds.dynamic.pipelineRevision,
            knownIds.dynamic.pipeline,
            '--format=json'
          ],
          { token: process.env.FACTORY_DYNAMIC_AUTH_TOKEN! }
        )
      );

      runId = data._id;
    });

    test('waits until run is finished', async () => {
      await expect(
        execute(
          [
            'get',
            'run',
            '--project-id',
            knownIds.dynamic.project,
            '--wait',
            runId,
            '--format=json'
          ],
          { token: process.env.FACTORY_DYNAMIC_AUTH_TOKEN! }
        ).then((value) => isArray(JSON.parse(value)))
      ).resolves.toBeTruthy();
    });
  });

  describe('job', () => {
    test('throws on missing arguments', async () => {
      await expect(
        execute(['run', 'job'], {
          token: process.env.FACTORY_DYNAMIC_AUTH_TOKEN!
        })
      ).rejects.toMatchSnapshot();
    });

    test('returns run object without --wait flag', async () => {
      await expect(
        execute(
          [
            'run',
            'job',
            '--project-id',
            knownIds.dynamic.project,
            '--var',
            '\'string_array_variable:["string", "array"]\'',
            '--var',
            '\'string_variable:"string"\'',
            '--var',
            "'number_array_variable:[100, 123]'",
            '--var',
            "'number_variable:123'",
            '--var',
            "'boolean_variable:true'",
            knownIds.dynamic.job,
            '--format=json'
          ],
          { token: process.env.FACTORY_DYNAMIC_AUTH_TOKEN! }
        ).then((value) => JSON.parse(value))
      ).resolves.toHaveProperty('status', 'Queued');
    });

    test('returns an array of events with --wait flag', async () => {
      await expect(
        execute(
          [
            'run',
            'job',
            '--project-id',
            knownIds.dynamic.project,
            '--wait',
            knownIds.dynamic.job,
            '--format=json'
          ],
          { token: process.env.FACTORY_DYNAMIC_AUTH_TOKEN! }
        ).then((value) => isArray(JSON.parse(value)))
      ).resolves.toBeTruthy();
    });

    test('correctly runs with variables', async () => {
      await expect(
        (async () => {
          const result = JSON.parse(
            await execute(
              [
                'run',
                'job',
                '--project-id',
                knownIds.dynamic.project,
                '--var',
                '\'string_array_variable:["string", "array"]\'',
                '--var',
                '\'string_variable:"string"\'',
                '--var',
                "'number_array_variable:[100, 123]'",
                '--var',
                "'number_variable:123'",
                '--var',
                "'boolean_variable:true'",
                '--wait',
                knownIds.dynamic.job,
                '--format=json'
              ],
              { token: process.env.FACTORY_DYNAMIC_AUTH_TOKEN! }
            )
          );

          if (!isArray(result)) {
            throw new Error('Expected array to be received!');
          }

          const ret = result.find(
            (entry) =>
              entry.message === 'Debug message' && entry.code === 'StepOutput'
          );
          if (isNil(ret)) {
            throw new Error('Expected debug step output to be presented!');
          }

          return JSON.parse(ret.details);
        })()
      ).resolves.toStrictEqual({
        string_array_variable: ['string', 'array'],
        string_variable: 'string',
        number_array_variable: [100, 123],
        number_variable: 123,
        boolean_variable: true
      });
    });
  });
});
