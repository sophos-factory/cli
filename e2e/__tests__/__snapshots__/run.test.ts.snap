// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`refactrctl run job throws on missing arguments 1`] = `
"refactrctl run job [job-id]

Executes specified job

Positionals:
  job-id  ID of the job to be run  [string] [default: read from stdin]

Options:
      --version             Show version number  [boolean]
  -v, --verbose             Print detailed output  [boolean]
      --format              Output format  [choices: "wide", "json", "yaml"] [default: "wide"]
      --filter              Filter output using JsonPath  [string]
      --address             Address of the Sophos Factory API server  [string] [default: FACTORY_ADDRESS environment variable if set, otherwise https://api.dev.factory.sophos.com/v1]
      --auth-token          Authentication token  [string] [default: FACTORY_AUTH_TOKEN environment variable]
      --help                Show help  [boolean]
      --wait                Wait until run execution is finished  [boolean]
      --suppress-events     Suppress run events during pipeline or job run  [boolean]
      --suppress-outputs    Suppress outputs during pipeline or job run  [boolean]
      --suppress-variables  Suppress variables during pipeline or job run  [boolean]
      --var                 Pipeline variable in \`key:value\` format, where value is JSON data or valid JSON object.  [string]
      --project-id          ID of the project containing the job  [string] [required]

Missing required argument: project-id
"
`;

exports[`refactrctl run pipeline throws on missing arguments 1`] = `
"refactrctl run pipeline [pipeline-id]

Executes specified pipeline

Positionals:
  pipeline-id  ID of the pipeline to be executed  [string] [default: read from stdin]

Options:
      --version             Show version number  [boolean]
  -v, --verbose             Print detailed output  [boolean]
      --format              Output format  [choices: "wide", "json", "yaml"] [default: "wide"]
      --filter              Filter output using JsonPath  [string]
      --address             Address of the Sophos Factory API server  [string] [default: FACTORY_ADDRESS environment variable if set, otherwise https://api.dev.factory.sophos.com/v1]
      --auth-token          Authentication token  [string] [default: FACTORY_AUTH_TOKEN environment variable]
      --help                Show help  [boolean]
      --wait                Wait until run execution is finished  [boolean]
      --suppress-events     Suppress run events during pipeline or job run  [boolean]
      --suppress-outputs    Suppress outputs during pipeline or job run  [boolean]
      --suppress-variables  Suppress variables during pipeline or job run  [boolean]
      --var                 Pipeline variable in \`key:value\` format, where value is JSON data or valid JSON object.  [string]
      --project-id          ID of the project containing the pipeline  [string] [required]
      --revision-id         ID of the pipeline revision to be executed  [string] [required]

Missing required arguments: project-id, revision-id
"
`;
